tests/test_brownian_interval.py

```diff
@@ -0,0 +1,315 @@
+
+        for shape_ in shapes:
+            assert shape_ == shape
+        for shape_ in A_shapes:
+            assert shape_ == A_shape
+
+
+@pytest.mark.parametrize("device", devices)
+@pytest.mark.parametrize("levy_area_approximation, return_U, return_A", _levy_returns())
+def test_determinism_simple(device, levy_area_approximation, return_U, return_A):
+    if device == gpu and not torch.cuda.is_available():
+        pytest.skip(msg="CUDA not available.")
+
+    ta, tb, bm = _setup(device, levy_area_approximation, (SMALL_BATCH_SIZE, D))
+    vals = [bm(ta, tb, return_U=return_U, return_A=return_A) for _ in range(REPS)]
+    for val in vals[1:]:
+        if torch.is_tensor(val):
+            val = (val,)
+        if torch.is_tensor(vals[0]):
+            val0 = (vals[0],)
+        else:
+            val0 = vals[0]
+        for v, v0 in zip(val, val0):
+            assert (v == v0).all()
+
+
+@pytest.mark.parametrize("device", devices)
+@pytest.mark.parametrize("levy_area_approximation, return_U, return_A", _levy_returns())
+def test_determinism_large(device, levy_area_approximation, return_U, return_A):
+    """
+    Tests that BrownianInterval deterministically produces the same results when queried at the same points.
+
+    We first of all query it at lots of points (larger than its internal cache), and then re-query at the same set of
+    points, and compare.
+    """
+    if device == gpu and not torch.cuda.is_available():
+        pytest.skip(msg="CUDA not available.")
+
+    ta, tb, bm = _setup(device, levy_area_approximation, (SMALL_BATCH_SIZE, D))
+    cache = {}
+    for _ in range(LARGE_REPS):
+        ta_ = torch.rand_like(ta)
+        tb_ = torch.rand_like(tb)
+        ta_, tb_ = min(ta_, tb_), max(ta_, tb_)
+        val = bm(ta_, tb_, return_U=return_U, return_A=return_A)
+        if torch.is_tensor(val):
+            val = (val,)
+        cache[ta_, tb_] = tuple(v.detach().clone() for v in val)
+
+    cache2 = {}
+    for ta_, tb_ in cache:
+        val = bm(ta_, tb_, return_U=return_U, return_A=return_A)
+        if torch.is_tensor(val):
+            val = (val,)
+        cache2[ta_, tb_] = tuple(v.detach().clone() for v in val)
+
+    for ta_, tb_ in cache:
+        for v1, v2 in zip(cache[ta_, tb_], cache2[ta_, tb_]):
+            assert (v1 == v2).all()
+
+
+@pytest.mark.parametrize("device", devices)
+@pytest.mark.parametrize("levy_area_approximation", ['none', 'space-time', 'davie', 'foster'])
+def test_normality_simple(device, levy_area_approximation):
+    if device == gpu and not torch.cuda.is_available():
+        pytest.skip(msg="CUDA not available.")
+
+    t0_, t1_ = 0.0, 1.0
+    t0, t1 = torch.tensor([t0_, t1_], device=device)
+    for _ in range(REPS):
+        W = torch.tensor(npr.randn(), device=device).repeat(LARGE_BATCH_SIZE)
+        bm = torchsde.BrownianInterval(t0=t0, t1=t1, W=W, levy_area_approximation=levy_area_approximation)
+
+        for _ in range(REPS):
```

Comment:
Have you run this test? It looks somewhat suspicious to me that it will pass. As you repeat the bm query, the conditional statistics would become different than just conditioning on the left most and right most endpoints (t0, t1 in this case), right?

Reply:
Yes, I've run it. It should work in terms of theory.
If I query the BM and then ignore my observation then I might as well have not queried it at all - distributionally (or in this case, over a batch of size 32000), it's the same.

If you prefer, this is tower law. For any one sample, sure, the conditional statistics change. Take the expectation (over a batch of size 32000) and the condition drops out.

Reply:
That makes sense, but I as run your branch on my local machine, these normality tests fail specifically. Am I missing anything?

Reply:
I'm getting pvals very much close to being 0.

Reply:
Ah, interestingly it looks like there's some nondeterminism involved. Sometimes I get all passes; sometimes I get a single failure mixed in as well. I'm not sure where that's coming from. (Sidenote: the GPU tests were never going to be deterministic because of nondeterminism in CUDA / nonassociativity of floating point arithmetic, but I'm also observing this on the CPU as well.)

Regardless, the pvals I observe seem pretty good; here's a sample from one run:
```
0.1831880731221287
0.7203067705273924
0.4005865963923665
0.9836020139630008
0.595396068278881
0.754736013617661
0.49089814175274626
0.8991110938461623
0.9471682808052452
0.06838178984312837
0.1917608973567274
0.10887317980591127
0.4471597979510218
0.2429775832482548
0.5234704439252269
0.9927171941269395
0.7041367966736902
0.6697926636524723
0.966971540959411
0.31859260562945524
0.49846075734446305
0.9633103124760822
0.25049223541296295
0.7333669455911738
0.8127956811078841
0.7822939021877175
0.8943048913302614
0.655253455597571
0.7993983708378245
0.14585232647852497
0.15457143178071975
0.8382653075582929
0.6702470486709494
0.4400739985187604
0.7388889139761349
0.10928069948262087
0.2908813124306524
0.12164793337455271
0.2304577968211593
0.6821789375137386
0.3738747784267529
0.15103241471340767
0.09241350508792956
0.13738289156372208
0.9355751354566113
0.19981707556298756
0.17567687104752305
0.875303717188714
0.6463672817110033
0.7680254937059569
0.9855047039836732
0.9096026698627325
0.47539405333075624
0.0021910005327920225
0.14046700256713318
0.31588096252673215
0.6771577262131989
0.9723814652857421
0.036046821301315646
0.3272079086543058
0.3236566425640489
0.4085607048407861
0.030045800063961514
0.8577012449476132
0.6006252116500315
0.33018822367454104
0.6688857325347278
0.8692185469427525
0.006884728052989697
0.09794983997900208
0.9295964081955631
```
I wouldn't be rejecting the null hypothesis as a result of these.

Can you double check you've got the latest version of the code, and share what kind of p-values you observe?

Reply:
Yes, I was running the on the latest commit. Here're some pvals I get from running the conditional normal test
```
0.5607581684785437
0.5919312036699521
0.014969262467999029
0.9024321662628461
0.48012136689587176
5.224794616863965e-78
0.5453968496808617
4.074461892756811e-20
2.121709516398075e-181
```

Reply:
This is on CPU btw.

Reply:
That's fascinating. Perhaps whatever PRNG is being used under the hood in PyTorch is different between us, and for whatever reason mine of better quality?
Can you try re-running on a different machine, preferably a different OS/architecture/etc. _(you know things are getting interesting when that's a suggestion)_ and see what happens?
Additionally, also try increasing the pool size and see if that changes things?

Reply:
Yeah I agree this is interesting. Switching machines and increasing pool size did not help. The tests are failing for the following environments for me
```
PyTorch version: 1.6.0
Is debug build: False
CUDA used to build PyTorch: None

OS: Mac OSX 10.15.4 (x86_64)
GCC version: Could not collect
Clang version: 11.0.3 (clang-1103.0.32.29)
CMake version: version 3.16.5

Python version: 3.7 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
```

```
PyTorch version: 1.6.0+cu101
Is debug build: False
CUDA used to build PyTorch: 10.1

OS: Debian GNU/Linux 9.12 (stretch) (x86_64)
GCC version: (Debian 6.3.0-18+deb9u1) 6.3.0 20170516
Clang version: Could not collect
CMake version: Could not collect

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.1.243
GPU models and configuration:
GPU 0: Tesla P100-PCIE-16GB
GPU 1: Tesla P100-PCIE-16GB
GPU 2: Tesla P100-PCIE-16GB
GPU 3: Tesla P100-PCIE-16GB

Nvidia driver version: 418.87.01
cuDNN version: Could not collect
```
Did you push the latest commits to your branch? Also could you share your system info?

Reply:
I guess in the short term, just to ensure the tests aren't too flaky, we should probably reduce REPS/increase batch size or altogether change/remove the tests here. 

Definitely worth coming back for another revisit.

Reply:
Yep, definitely using the latest version.

Things seem fine for me with either Python 3.7 or Python 3.8 on Linux.

What versions of the relevant libraries do you have? I could believe that a difference in scipy or mkl_random could affect things.

My environments:

Environment 1:
```
PyTorch version: 1.6.0
Is debug build: False
CUDA used to build PyTorch: 10.2

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.10.2

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti
GPU 2: Quadro GP100
GPU 3: Quadro GP100

Nvidia driver version: 450.51.06
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip3] numpy==1.19.1
[pip3] torch==1.6.0
[pip3] torchsde==0.1.2
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               10.2.89              hfd86e86_1
[conda] mkl                       2020.1                      217
[conda] mkl-service               2.3.0            py37he904b0f_0
[conda] mkl_fft                   1.1.0            py37h23d657b_0
[conda] mkl_random                1.1.1            py37h0573a6f_0
[conda] numpy                     1.19.1           py37hbc911f0_0
[conda] numpy-base                1.19.1           py37hfa32c7d_0
[conda] pytorch                   1.6.0           py3.7_cuda10.2.89_cudnn7.6.5_0    pytorch
[conda] torchsde                  0.1.2                     dev_0    <develop>
```

Environment 2:
```
PyTorch version: 1.6.0
Is debug build: False
CUDA used to build PyTorch: 10.2

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.10.2

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti
GPU 2: Quadro GP100
GPU 3: Quadro GP100

Nvidia driver version: 450.51.06
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip3] numpy==1.19.1
[pip3] torch==1.6.0
[pip3] torchsde==0.1.2
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               10.2.89              hfd86e86_1
[conda] mkl                       2020.1                      217
[conda] mkl-service               2.3.0            py38he904b0f_0
[conda] mkl_fft                   1.1.0            py38h23d657b_0
[conda] mkl_random                1.1.1            py38h0573a6f_0
[conda] numpy                     1.19.1           py38hbc911f0_0
[conda] numpy-base                1.19.1           py38hfa32c7d_0
[conda] pytorch                   1.6.0           py3.8_cuda10.2.89_cudnn7.6.5_0    pytorch
[conda] torchsde                  0.1.2                     dev_0    <develop>
```

Reply:
Regarding flakiness of tests - do you want to make the adjustments for that, as I have no way of testing what makes it not flakey! I'll leave this branch alone for you to make changes on.

Unrelatedly, is there anything you think I should particularly look at next? There's bits and bobs I can do but I think I've handled all the glaring items on my to-do list.

Reply:
Oh, and it seems like I didn't even list my own version of scipy! For 3.7, 3.8 respectively:
```
scipy                     1.5.0            py37h0b6359f_0
scipy                     1.5.0            py38h0b6359f_0
```

Reply:
I upgraded all the packages to have the same versions as you do, and the problem didn't go away. 

Could you double check if you get mean=0 and std=1 for the test [here](https://github.com/google-research/torchsde/blob/cae4bc2dd9b4ab41be8a3788dccba3437915c0e3/tests/test_brownian_interval.py#L186)? 

I'm seeing things fail systematically for `foster`/`david`/`space-time`.

I'm sorry if this gets grueling, but if things fail here or aren't stable enough, I'm pretty sure they'll fail even more miserably when we plug everything into the sdeint_adjoint code, and frankly debugging a `torch.nn.Function` can get annoying.

Reply:
A side note is that when I was testing bpath and btree, I noticed that things can get flaky when the time points are very close, so that's why I added an `eps` guard to prevent that from happening.

Reply:
> 
> 
> I upgraded all the packages to have the same versions as you do, and the problem didn't go away.
> 
> Could you double check if you get mean=0 and std=1 for the test [here](https://github.com/google-research/torchsde/blob/cae4bc2dd9b4ab41be8a3788dccba3437915c0e3/tests/test_brownian_interval.py#L186)?
> 
> I'm seeing things fail systematically for `foster`/`david`/`space-time`.
> 
> I'm sorry if this gets grueling, but if things fail here or aren't stable enough, I'm pretty sure they'll fail even more miserably when we plug everything into the sdeint_adjoint code, and frankly debugging a `torch.nn.Function` can get annoying.

Adding `print(torch.std_mean(torch.tensor(rescaled_sample)))` to the test prints out:
```
(tensor(1.0025), tensor(0.0006))
(tensor(1.0014), tensor(-0.0026))
(tensor(1.0000), tensor(-0.0006))
(tensor(1.0020), tensor(-0.0018))
(tensor(1.0009), tensor(-0.0034))
(tensor(1.0010), tensor(0.0022))
(tensor(0.9997), tensor(0.0007))
(tensor(0.9977), tensor(0.0029))
(tensor(0.9999), tensor(0.0016))
(tensor(0.9971), tensor(0.0017))
(tensor(0.9979), tensor(0.0020))
(tensor(0.9994), tensor(-0.0053))
(tensor(1.0013), tensor(-0.0019))
(tensor(1.0012), tensor(0.0018))
(tensor(0.9993), tensor(-0.0008))
(tensor(1.0021), tensor(-0.0033))
(tensor(0.9992), tensor(-0.0011))
(tensor(1.0021), tensor(0.0064))
(tensor(1.0039), tensor(-0.0108))
(tensor(0.9979), tensor(0.0018))
(tensor(0.9984), tensor(-0.0030))
(tensor(0.9970), tensor(0.0034))
(tensor(0.9989), tensor(-0.0063))
```
which LGTM.

And don't worry, I appreciate that "it works on my machine" isn't going to cut it.

Are you saying that `none` seems to work? (That uses a different algorithm to the other three, so I can believe it.)

Okay, to keep digging in: when you run
```

Reply:
> 
> 
> A side note is that when I was testing bpath and btree, I noticed that things can get flaky when the time points are very close, so that's why I added an `eps` guard to prevent that from happening.

What kind of flakiness do you observe? I didn't _too_ much thought into numerical stability whilst I was writing things (although I did a bit). If possible I'd rather rewrite the computations if they're unstable rather than add another feature to the mix.

Reply:
> do you get 2029077620 as the valueI believe SeedSequence is a fixed implementation that should be the same for both of us, so it's a good place to start.

I get 
```771466205```

> torch.randn(64, generator=torch.Generator().manual_seed(seed))

I get 
```
tensor([ 0.7314, -0.7575,  0.5196, -0.2246, -1.1200,  1.2430,  1.2759,  0.0239,
        -0.1964, -0.4379, -0.8360, -0.7990, -0.7343,  0.7382,  2.3253, -0.0809,
        -1.3725, -0.4227, -0.3373, -1.7258, -0.8709,  0.1395,  1.2477, -0.1724,
         0.3683, -0.7075,  1.1928, -0.2962,  0.8203, -1.7916, -0.7161,  0.1629,
        -0.3906,  0.3133,  0.0258, -0.3784,  0.3754, -0.3247,  1.1662, -0.7479,
        -1.9184,  1.1434,  1.7734, -0.0771,  1.1552, -0.4181,  0.6999,  0.5334,
         0.7735,  0.2979, -1.7747, -0.4775,  0.2538,  0.2146, -1.2433,  1.2250,
         0.6106, -2.3741,  0.6630, -1.3142,  0.8098,  0.8615, -1.0300, -1.7857])
```

On a broader note, I don't think these check are indicative of much. 

> What kind of flakiness do you observe? I didn't too much thought into numerical stability whilst I was writing things (although I did a bit). If possible I'd rather rewrite the computations if they're unstable rather than add another feature to the mix.

So when the query time is very close to one side for Bbridge, there could be a problem of adding large numbers to small numbers which is not so great in the context of floating point computation.

Reply:
Cloning a fresh repo and reinstalling your branch, with `print(torch.std_mean(torch.tensor(rescaled_sample)))`, I'm getting

```
(tensor(0.9987), tensor(0.0003))
(tensor(0.9980), tensor(-0.0024))
(tensor(0.9973), tensor(0.0036))
(tensor(0.6418), tensor(-0.0011))
(tensor(0.6476), tensor(0.0030))
(tensor(0.8856), tensor(5.2828e-05))
```

Reply:
Au contraire, I think it's identified the exact problem. You're using an old version of numpy; upgrade to the latest 1.19.1.

Reply:
OK, it's working now. Could you send in another PR bumping up the version in `setup.py`?